blaze/compute/core.py

```diff
@@ -446,12 +446,13 @@ def compute(expr, d, **kwargs):
     optimize_ = kwargs.get('optimize', optimize)
     pre_compute_ = kwargs.get('pre_compute', pre_compute)
     post_compute_ = kwargs.get('post_compute', post_compute)
-
     expr2, d2 = swap_resources_into_scope(expr, d)
     if pre_compute_:
-        d3 = dict([(e, pre_compute_(expr2, dat, **kwargs))
```

Comment:
@mrocklin Are these the correct arguments to `pre_compute`? For expressions with multiple leaves like `Join` this does the wrong thing because it calls `into` with the dshape of the first leaf that it finds. According to the pipeline docs, `pre_compute` operates on the leaves of the expression, which is more inline with @llllllllll change here to `pre_compute` each expression in scope.